### PR TITLE
planner: add functional dependency doc

### LINKED ISSUE #29766: Refine only-full-group-by checker
## Enhancement

TiDB's only-full-group-by checker is not so compatible with MySQL. The incompatibility brings trouble to users and the current workaround is to remove `ONLY_FULL_GROUP_BY` from `sql_mode`. We need to make only-full-group-by checker more compatible with MySQL.

Here are the issues related to only-full-group-by checker:
- https://github.com/pingcap/tidb/issues/29418
- https://github.com/pingcap/tidb/issues/27723
- https://github.com/pingcap/tidb/issues/27539
- https://github.com/pingcap/tidb/issues/26576
- https://github.com/pingcap/tidb/issues/25199
- https://github.com/pingcap/tidb/issues/25196
- https://github.com/pingcap/tidb/issues/22694
- https://github.com/pingcap/tidb/issues/22301
- https://github.com/pingcap/tidb/issues/25089
- https://github.com/pingcap/tidb/issues/21848
- https://github.com/pingcap/tidb/issues/30024

One cause for some issues is that only-full-group-by checker is not smart enough to infer some functional dependency(See https://github.com/pingcap/tidb/issues/14580 for details).

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #29766

Problem Summary:

### What is changed and how it works?
Motivation Issue: https://github.com/pingcap/tidb/issues/29766
add doc of functional dependency

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-03-16-functional-dependency.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+For outer join, we need to consider more, because outer join will choose the connection method based on the position of the join conditions and the presence of the join key.
+
+* only inner side condition           --- apply inner filter after the cartesian product is made, which means no supplied-null row will be appended on the inner side.
+* has outer condition or join key     --- apply all of this filter as matching judgement on the process of cartesian product, which means supplied-null row will be appended on the inner side when matching is failed.
+
+both cases will keep all rows from the outer side, but the time to apply the filter depends on where the conditions come from, which leading whether append the null rows on the inner side. So what's the big deal between this two? the former one can be seen as a inner join, but the later can only keep the FD from the outer side join, inner side's FD and filter FD has many complicated rules to infer with, we are not going to expand all of this here.
```
Comment (time-and-fate) [inline docs/design/2022-03-16-functional-dependency.md:136]: This paragraph is not very easy to understand.

Comment (AilinKid) [reply docs/design/2022-03-16-functional-dependency.md:136]: make sense, I have rewritten.

File: docs/design/2022-03-16-functional-dependency.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+For outer join, we need to consider more, because outer join will choose the connection method based on the position of the join conditions and the presence of the join key.
+
+* only inner side condition           --- apply inner filter after the cartesian product is made, which means no supplied-null row will be appended on the inner side if any row is preserved; or all rows are supplied-null if all rows are filtered out.
+* has outer condition or join key     --- apply all of this filter as matching judgement on the process of cartesian product, which means supplied-null row will be appended on the inner side when matching is failed.
+
+Although both cases will keep all rows from the outer side, but how to derive FDs from both side depends on where the conditions come from, what the columns used by the conditions are, and what the attributes the conditions can provide. We can treat the former one as an inner join, but for the later one, we can only keep the FD from the outer table of join, and inner side's FD and filter FD has many complicated rules to infer with, we are not going to expand all of this here.
```
Comment (time-and-fate) [inline docs/design/2022-03-16-functional-dependency.md:136]:
```suggestion
Although both cases will keep all rows from the outer side, but how to derive FDs from both sides depends on where the conditions come from, what the columns used by the conditions are, and what the attributes the conditions can provide. We can treat the former one as an inner join, but for the latter one, we can only keep the FD from the outer table of join, and inner side's FD and filter FD has many complicated rules to infer with, we are not going to expand all of this here.
```

File: docs/design/2022-03-16-functional-dependency.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+#### How to integrate functional dependency with TiDB arch
+
+Where to put this stuff in and when should the FD collection work be done? In FD logic, the relationship between columns is not limited to base col, but also between expressions, even sub-queries. We need to be able to identify sameness between cols, expressions and even better to have a unique id to tag them. When joining different tables, there are maybe some columns with the same name from difference source, we should still utilize this unified unique id to identify and distinguish them. Obviously, the expression tree after ast rewrite has this convenient functionality, and plus we need to utilize FD functionality to check validation when building a logical plan, so eventually we can only perform FD collection while building the logical plan stage.
```
Comment (time-and-fate) [inline docs/design/2022-03-16-functional-dependency.md:99]:
```suggestion
Where to put this stuff in and when should the FD collection work be done? In FD logic, the relationship between columns is not limited to base col, but also between expressions, even sub-queries. We need to be able to identify sameness between cols, expressions and even better to have a unique id to tag them. When joining different tables, there are maybe some columns with the same name from different sources, we should still utilize this unified unique id to identify and distinguish them. Obviously, the expression tree after ast rewrite has this convenient functionality, and plus we need to utilize FD functionality to check validation when building a logical plan, so eventually we can only perform FD collection while building the logical plan stage.
```

**Review by time-and-fate — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**